net: lwm2m: rework error handing in content formaters

rlubos opened the pull request:
This PR introduces a set of changes to improve error detection in content encoders/decoders;
* Modify the encoders/decoders generic API, to return `ssize_t` instead of `size_t`, therefore allowing to report error conditions (typically -ENOMEM for encoders when we run out of memory in a packer, or -ENODATA for decoders when there is no expected data in received packet),
* Align all of the existing encoders/decoders with the API changes, adding additional error checks where appropriate,
* Add a set of unit tests for each content type encoder/decoder,
* And finally fix a couple of minor bug detected by the aforementioned tests.

Fixes #41996.

rlubos committed:
net: lwm2m: Allow to return error codes by content encoders/decoders

Rework the content encoder/decoder API to allow to return negative
values as well. This allows a proper error reporting and error handling,
as it's now possible to differentiate when there is no content to write
(retuned 0) or and error occured, and further processing should be
aborted.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Improve link-format content encoder error handling

It's now possible to return an error code instead of 0 where
appropriate.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Improve plain-text content encoder/decoder error handling

It's now possible to return an error code instead of 0 where
appropriate.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Improve TLV content encoder/decoder error handling

It's now possible to return an error code instead of 0 where
appropriate.

Make `buf_read()` return -ENODATA instead of -ENOMEM if there's not
enough data in the packet to serve the request, which is more meaningful
for this scenario.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Improve JSON content encoder/decoder error handling

It's now possible to return an error code instead of 0 where
appropriate.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Improve content writers error handling in engine

Verify the restult of the content writer functions and return an error
if content writer fails to read/write field in the message.

This solves an issue when for example malformed packet was sent to the
server if the payload did not fit in the message.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: lwm2m: Add link-format content writer tests

Add unit tests for LwM2M link-format content encoder/decoder.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: lwm2m: Add plain text content writer tests

Add unit tests for LwM2M plain text conent encoder/decoder.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Fix lwm2m_atof for integer input

In case a string with an integer was provided to the function (no
decimal point), the function did not update the output pointer value.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: lwm2m: Add OMA TLV content writer tests

Add unit tests for LwM2M OMA TLV conent encoder/decoder.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Fix OMA TLV objlnk encoding

At some point OMA TLV integer encoding was optimised to use the smallest
size possible. This broke the objlnk encoding, which internally used
`put_s32()`, but should always be 4 bytes long. In case the result
32-bit integer would fit into 16 or 8 bytes, it'd get optimized.

Fix this, by creating and writing TLV manually for objlnk instead of
relying on int32 encoder.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
tests: net: lwm2m: Add JSON content writer tests

Add unit tests for LwM2M JSON conent encoder/decoder.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
CC @jheiskan81 @VeijoPesonen @SeppoTakalo feel free to review

rlubos commented:
@jukkar My bad, I didn't notice `ssize_t` is limited to -1 only by the specification. I agree we should not use it to return other negative values to avoid confision and I'm not really a fan of using `errno` for this use case.

> Just pondering here that would it make sense to use normal int value to tell error (<0), and 0 as success and then have a separate size_t value returned as a parameter, what do you think?

I think we could use the `int` for the return value, I just wonder do we really need that extra parameter for returning the size written? Those functions work with datagrams, which size is limited to 64K so in theory there's no way we'll overflow `int` if we return the number of bytes written/read directly through the return value (we could use `int32_t` just to be 100% sure).

jukkar commented:
> I think we could use the `int` for the return value, I just wonder do we really need that extra parameter for returning the size written? Those functions work with datagrams, which size is limited to 64K so in theory there's no way we'll overflow `int` if we return the number of bytes written/read directly through the return value (we could use `int32_t` just to be 100% sure).

In that case, we could just return int as it can handle all the values returned.

rlubos committed:
net: lwm2m: Fix lwm2m_parse_peerinfo misuse

`lwm2m_parse_peerinfo()` expects a boolean value, not a pointer.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: lwm2m: Prevent snprintk warnings on different platforms

Use explicit casting to long long within `snprintk()` and logger
functions to prevent compiler warnings with different
platforms/toolchins (as 64-bit integer can be either represented
as ld or lld depending on platform).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
@jukkar I've updated the APIs to use `int` instead. Additionally added a couple of commits addressing nits found by the CI.

jheiskan81 reviewed (approved):
Looks good for me.

jukkar reviewed (approved):
LGTM

rlubos commented:
@carlescufi Any chance to get this in before code freeze?

VeijoPesonen commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -926,7 +924,7 @@ int do_write_op_tlv(struct lwm2m_message *msg)
 		 */
 		len = oma_tlv_get(&tlv, &msg->in, true);
-		if (len == 0) {
+		if (len < 0) {
```
Empty data is ok?

rlubos replied:
For TLV there's never a case where there is no data (i. e. 0 is returned) - either there is at least TLV header (and thus the return value is non-zero positive), or an error code is returned.

VeijoPesonen reviewed (approved):
One question - otherwise looks good to me.

SeppoTakalo commented on subsys/net/lib/lwm2m/lwm2m_rw_plain_text.c:
```diff
@@ -219,20 +224,20 @@ static size_t get_s32(struct lwm2m_input_context *in, int32_t *value)
 
 	if (in_len > buflen) {
-		/* TODO: generate warning? */
+		LOG_WRN("Buffer too small to accommodate string, truncating");
```
Should this be `LOG_ERR()` as we actually failed to generate the correct content.

rlubos replied:
I can increse log severity level, it might make sense.

rlubos commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -3160,12 +3182,20 @@ int lwm2m_perform_read_op(struct lwm2m_message *msg, uint16_t content_format)
+					/* No point continuing if there's no
+					 * memory left in a message.
+					 */
+					return ret;
```
Allowing to continue when `ENOMEM` is reported would require further rework fo the content writers - TLV or JSON will insert appropriate metadata in the packet first (TLV header, JSON tokens). If we fail to write the payload afterward, we still end up with a malformed packed (it should be required from content writers to undo any changes done to the packet). Therefore for now I've decided to treat `ENOMEM` as fatal.

Another thing is whether handling the requests partially is ok - responding with an error message could hint the server that it should query each resource individually, otherwise the event would go unnoticed, resulting in outdated reasource values on the server side.

petterssonandreas commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2545,72 +2560,69 @@ int lwm2m_write_handler(struct lwm2m_engine_obj_inst *obj_inst,
-			engine_get_string(&msg->in, write_buf, write_buf_len);
+			ret = engine_get_string(&msg->in, write_buf,
+						write_buf_len);
 			len = strlen((char *)write_buf);
```
Isn't this potentially bad? If `engine_get_string` produced an error, which we now can see by checking `ret`, we have no idea what `write_buf` actually contains right? Of course, this is only a read, but it could potentially read indefinitely, right, since we have no way of knowing that the string is terminated?

petterssonandreas commented on subsys/net/lib/lwm2m/lwm2m_engine.c:
```diff
@@ -2545,72 +2560,69 @@ int lwm2m_write_handler(struct lwm2m_engine_obj_inst *obj_inst,
 		case LWM2M_RES_TYPE_TIME:
-			engine_get_s64(&msg->in, &temp64);
+			ret = engine_get_s64(&msg->in, &temp64);
 			*(uint32_t *)write_buf = temp64;
```
This seems bad, do we really want to write the data to `write_buf` without checking `ret` first to make sure we didn't get any errors? Similarly for all the other types.

rlubos replied:
You're right, this isn't right to write the value in case of error, I've added additional error check before write here and in other cases.

petterssonandreas commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -537,48 +526,63 @@ static size_t put_float(struct lwm2m_output_context *out,
+	size = oma_tlv_get(&tlv, in, false);
+	if (size < 0) {
+		return size;
+	}
```
I prefer this early return, because that makes it clear what we do for errors. But we don't use early return in the other get functions, so maybe convert all to early returns?

rlubos replied:
I've updated other functions to return early as well, instead of nesting code within if.

petterssonandreas commented on subsys/net/lib/lwm2m/lwm2m_rw_oma_tlv.c:
```diff
@@ -537,48 +526,63 @@ static size_t put_float(struct lwm2m_output_context *out,
+	}
 
 	*value = 0;
 	if (size > 0) {
```
Do we need this if? Or should the if above include 0 also?

Do we really want to set `*value = 0;` even if `size` is 0?

rlubos replied:
Yeah, this check seems redundant now, I"ve remoted that if. `size` shouldn't be 0 for TLVs, as it's always expected to have at least the TLV header - so it'd be either a positive value, or a negative error code.
Finally, after a second check it doesn't seem to be needed to pre-initialize the returned  value to 0, the function shouldn't write anything in case of error.

petterssonandreas reviewed (commented):
I am not too familiar with the code, but I (quite quickly) reviewed the parts that I addressed in https://github.com/zephyrproject-rtos/zephyr/issues/41996, but I found some issues that I think should be addressed.

rlubos commented:
@petterssonandreas Thank you for the feedback, I've addressed your comments.

carlescufi commented:
This PR fixes #41996, so merging it in the 3.0.0 window.
